[PLUGIN-1715] Implement retry feature for BigQuery Execute Plugin

psainics opened the pull request:
## Implement retry feature for BigQuery Execute Plugin 

Jira : [PLUGIN-1715](https://cdap.atlassian.net/browse/PLUGIN-1715)

- In general internal errors should be retried.
- Do an exponential back-off retry on errors with reason `jobBackendError` , `jobInternalError`.
- Backoff requirements (exponential from 1 to 32 seconds with a multiplier of 2)


### Code Changes
- Add a new set that contains all reason we should retry on.
- A new custom exception that is raised when the error should be retried.
- A new function to generate the exponential back- off retry policy  
- A new dependency is added `dev.failsafe` that handles the retry logic based on the exception raised.

### Unit Tests
- Use mocks to test the behavior of the retry.

[PLUGIN-1715]: https://cdap.atlassian.net/browse/PLUGIN-1715?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

tivv commented on src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java:
```diff
@@ -43,16 +46,20 @@
 import io.cdap.plugin.gcp.bigquery.util.BigQueryUtil;
 import io.cdap.plugin.gcp.common.CmekUtils;
 import io.cdap.plugin.gcp.common.GCPUtils;
+import org.jetbrains.annotations.TestOnly;
```
Don't use jetbrains annotations in production code. This may bring unnessesary dependencies

fernst replied:
If you want to mark a method as visible only for the purposes of testing, use `@VisibleForTesting` which is already used elsewhere in the codebase.

psainics replied:
Removed the jetbrains annotation,  using `@ VisibleForTesting`

tivv commented on src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java:
```diff
@@ -69,8 +76,13 @@ public final class BigQueryExecute extends AbstractBigQueryAction {
 
-  private Config config;
+  @TestOnly
+  public BigQueryExecute(Config config) {
```
Usually package-only access is enough for testing

tivv replied:
Adding an explicit config would effectively remove a no-argument constructor. Which constructor will be used for the production run? If this one, it should not be marked as TestOnly

psainics replied:
Made this package only.

tivv commented on src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java:
```diff
@@ -69,8 +76,13 @@ public final class BigQueryExecute extends AbstractBigQueryAction {
   public static final String NAME = "BigQueryExecute";
   private static final String RECORDS_PROCESSED = "records.processed";
+  private final Config config;
+  private static final Set<String> retryOnReason = ImmutableSet.of("jobBackendError", "jobInternalError");
```
All constant should be uppercased

fernst replied:
Also the strings here should be in their own constants.

psainics replied:
Updated !

tivv commented on src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java:
```diff
@@ -129,19 +138,62 @@ public void run(ActionContext context) throws Exception {
+    Duration initialRetryDuration = Duration.ofSeconds(1);
+    Duration maxRetryDuration = Duration.ofSeconds(32);
+    int multiplier = 2;
+    int maxRetryCount = 5;
```
All of this should be configurable. Retries must be disableable

fernst replied:
Agree. Please add a `Retry Configuration` section in the UI and allow the users to fine-tune this behavior.

vikasrathee-cs replied:
@tivv @fernst It was mentioned previously on the bug ticket also that Bigquery SLA is for 32 seconds and that is the reason we have added these fix properties to go upto that limit. Same was confirmed in design doc also.
Can you please confirm in design doc also. Shared with you.
Also have a look at this. https://cloud.google.com/bigquery/sla#:~:text=%22-,Back%2Doff%20Requirements,-%22%C2%A0means%2C%20when

tivv replied:
Few points here:

-  Customer may not wish to wait so long. Yep, BQ may recover, but it's also possible that it's too long for given customer
- You are changing the behaviour. We must have an option to disable this behavior
- SLAs are statistical. They may still be violated
- SLAs are point in time. They may change and we need to be able to adjust the bahvior without code change

psainics replied:
I have added new UI field to turn this behavior off (default) and let user configure the parameters.

<img width="482" alt="image" src="https://github.com/data-integrations/google-cloud/assets/122770897/d9f711df-b6d9-4e66-82cf-d4004329f389">

tivv commented on src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java:
```diff
@@ -129,19 +138,62 @@ public void run(ActionContext context) throws Exception {
-      throw new RuntimeException(queryJob.getStatus().getExecutionErrors().toString());
+      LOG.error("The query job {} failed. Error: {}", jobId.getJob(), queryJob.getStatus().getError());
+      if (retryOnReason.contains(queryJob.getStatus().getError().getReason())) {
+        throw new BigQueryJobExecutionException(queryJob.getStatus().getError().getMessage());
```
It's a checked exception. What would it be converted to when we are out of retries?

tivv commented on src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  @Test
+  public void testExecuteQueryWithExponentialBackoffSuccess() throws InterruptedException {
```
You probably want immediate retries, otherwise this test would be waiting unnesessary

psainics replied:
Fixed !

tivv commented on src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java:
```diff
@@ -0,0 +1,113 @@
+    bq.executeQueryWithExponentialBackoff(bigQuery, queryJobConfiguration, context);
+  }
+
+  @Test(expected = java.lang.RuntimeException.class)
```
Why are you expecting just a RuntimeException here? It that it should go through retries and give some reasonable error

psainics replied:
Updated !

fernst reviewed (commented):
Agree with Vitalii's suggestions here. Please address.

tivv commented on src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java:
```diff
@@ -129,19 +143,65 @@ public void run(ActionContext context) throws Exception {
+      Failsafe.with(getRetryPolicy()).run(() -> executeQuery(bigQuery, queryConfig, context));
+    } catch (RuntimeException e) {
+      LOG.error("Failed to execute BigQuery job. Error: {}", e.getMessage());
+      throw new BigQueryJobExecutionException("Failed to execute BigQuery job. Reason: " + e.getMessage());
```
This way you are loosing all the additional information from original BigQueryException stack trace. There are only some case where you would want to hide all details of the underlying exception by using e.getMessage instead of attaching a root cause. I don't see why would we want to have it here. Same applies to the catch clause in `executeQuery`

psainics replied:
Updated !

psainics commented:
Changes Squashed !

tivv commented on src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java:
```diff
@@ -0,0 +1,198 @@
+
+  }
+
+  @Test(expected = BigQueryJobExecutionException.class)
```
I am still a bit confused what final sequence of exceptions would it be and, what's important, what would be a message to user. Could you please use assertThrows and validate the message as described here: https://www.baeldung.com/junit-assert-exception ? I am wondering if we would need to unwrap those exceptions to prevent unneeded layers / cluttered message

psainics replied:
The user will always see a `BigQueryJobExecutionException`

There are 3 cases for this to occur
1. Retry is not possible as the error reason is not in reason list, the user get the message `Failed to execute BigQuery job. Reason: {Reason from that exception}`

2. Retry was tried but failed , the user will get the message ~`Failed to execute BigQuery job.  io.cdap.plugin.gcp.bigquery.exception.BigQueryJobExecutionException`~ `Failed to execute BigQuery job. Reason: Retries exhausted.`
- We could provide a different message here if required, as the exception class will be  `FailsafeException` for which an additional catch could be added .

3. Someting else fails and another exception is thrown , `Failed to execute BigQuery job. Reason: {Reason from that exception}`

tivv replied:
I am curious about the text of the message. I want to avoid something like  `Failed to execute BigQuery job. Reason: RuntimeException: BigQueryJobExecutionException: reason`

psainics replied:
I have updated to assert throw, and made it more clear what will the end user see. in all cases.

fernst reviewed (approved):
LGTM

tivv commented on src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java:
```diff
@@ -0,0 +1,210 @@
+      bq.executeQueryWithExponentialBackoff(bigQuery, queryJobConfiguration, context);
+    });
+    String actualMessage = exception.getMessage();
+    Assert.assertEquals("Failed to execute BigQuery job. Reason: " + mockErrorMessageNoRetry, actualMessage);
```
So, basically we say "Failed to execute BigQuery job. Reason: Job execution failed with error: $error" This is what I want to avoid as it's saying same thing twice

tivv replied:
May be in top level catch we should just find the correct cause and rethrow in instead of wrapping it again

psainics replied:
Sure, i have removed the wrapping of error message.

tivv replied:
But this way you are loosing the underlying error. You should either: 
 - do visa versa and pop up only the underlying error message "Job execution failed with error: $error"
  - prepend it with some text that adds value, e.g. "Job retries failed after X tries. Last try failure: Job execution failed with error: $error".

psainics replied:
Yup I am doing the 1st one , by removing wrapping I ment removing the extra messages string that was present before

tivv replied:
But you are hiding the underlying "$error", so the only thing that we can see is that retries exhausted without much information about the underlying error.

psainics replied:
As i already have logging when nth retry is made or the policy is out of retry should i just pop the error and throw that ?

something like this
```java
try {
      Failsafe.with(getRetryPolicy()).run(() -> executeQuery(bigQuery, queryConfig, context));
    } catch (FailsafeException e) {
      throw e.getCause();
    } 
``` 
Also i could drop the `catch (RuntimeException e)`

tivv replied:
Sure, that's option #1 in the list above

tivv replied:
I would check if e.getCause is not null as generally not every failure is an exception

psainics replied:
🤔 
```java
if (e.getCause() != null) {
        throw e.getCause();  
} 
  throw e;
```

psainics commented:
Thanks @tivv , i have made the changes and squashed them , hopefully you can approve this now.
I will get this QA tested before merging  👍

itsankit-google commented on widgets/BigQueryExecute-action.json:
```diff
@@ -193,6 +193,63 @@
+              "value": "false",
+              "label": "NO"
+            },
+            "default": "false"
```
shouldn't this be set to `true` by default, @tivv @fernst wdyt??
I thought aim is to add some resilency to the plugin to reduce intermittent failures but if the behavior is `false` by default, users will need to enable it explicitly and modify every pipeline.

psainics replied:
Your point makes a lot of sense. Setting this behavior to true by default aligns with the goal of enhancing plugin resilience and reducing intermittent failures.

tivv replied:
So, we usually have two things:
 * Default for UI. This is what we want default experience to be. In this case it should be "true"
 * Default for backend (when option is not set). This is usually how things worked before to ensure backward compatibility unless we have explicit OK from product to break backward compatibility

psainics replied:
@itsankit-google we had a discussion with deepinder, the UI fields should be hidden and true by default !

psainics commented:
New changes include 
- Reverted changes in `docs/BigQueryExecute-action.md` as the UI fields are hidden.
- Removed filter `retryOnBackendError` with condition `retryOnBackendError == true` as this will not be required now.
- Made all the ui fields `hidden`
- Retry On Backend Error is `TRUE` by default , on widget and backed.

psainics commented:
Resolved conflicts caused by PR: https://github.com/data-integrations/google-cloud/pull/1344 in `bigquery/action/BigQueryExecute.java`
